File: src/main/java/no/nav/syfo/util/MDCOperations.java
package no.nav.syfo.util;

import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;

import java.security.SecureRandom;

/**
 * Utility-klasse for kommunikasjon med MDC.
 */
@Slf4j
public final class MDCOperations {
    public static final String MDC_CALL_ID = "callId";
    public static final String MDC_USER_ID = "userId";
    public static final String MDC_CONSUMER_ID = "consumerId";

    private static final SecureRandom RANDOM = new SecureRandom();

    private MDCOperations() {
    }

    public static String generateCallId() {
        int randomNr = getRandomNumber();
        long systemTime = getSystemTime();

        return "CallId_" + systemTime + "_" + randomNr;
    }

    public static String getFromMDC(String key) {
        String value = MDC.get(key);
        log.debug("Getting key: " + key + " from MDC with value: " + value);
        return value;
    }

    public static void putToMDC(String key, String value) {
        log.debug("Putting value: " + value + " on MDC with key: " + key);
        MDC.put(key, value);
    }

    public static void remove(String key) {
        log.debug("Removing key: " + key);
        MDC.remove(key);
    }

    private static int getRandomNumber() {
        return RANDOM.nextInt(Integer.MAX_VALUE);
    }

    private static long getSystemTime() {
        return System.currentTimeMillis();
    }
}

File: src/main/kotlin/no/nav/syfo/util/MDCOperations.kt
package no.nav.syfo.util

import java.security.SecureRandom
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.slf4j.MDC

class MDCOperations {
    companion object {
        var log = logger<MDCOperations>()
        const val MDC_CALL_ID = "callId"
        const val MDC_USER_ID = "userId"
        const val MDC_CONSUMER_ID = "consumerId"
        private val RANDOM = SecureRandom()

        fun generateCallId(): String = "CallId_${getRandomNumber()}_${getSystemTime()}"

        fun getFromMDC(key: String): String? {
            val value = MDC.get(key)
            log.debug("Getting key: $key from MDC with value: $value")
            return value
        }

        fun putToMDC(key: String, value: String) {
            log.debug("Putting value: $value on MDC with key: $key")
            MDC.put(key, value)
        }

        fun remove(key: String) {
            log.debug("Removing key: $key")
            MDC.remove(key)
        }

        private fun getRandomNumber(): Int = RANDOM.nextInt(Int.MAX_VALUE)
        private fun getSystemTime(): Long = System.currentTimeMillis()
    }
}

inline fun <reified T> logger(): Logger {
    return LoggerFactory.getLogger(T::class.java)
}
